### PR TITLE
Adding support for UEFI emulation

### PR DESCRIPTION
**Pull request by liba2k**

### CODE REVIEW

**Comment by aquynh** [comment]
Awesome, but can you please fix the conflicts with the latest code?

File: qiling/os/uefi/uefi.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+            self.map_and_load(self.ql.path)
+
+            self.ql.nprint("[+] Done with loading %s" % self.ql.path)
+            self.filepath = b"D:\\" + bytes(self.ql.path.replace("/", "\\"), "utf-8")
+
+        elif self.ql.shellcoder:
+            # setup stack memory
+            self.ql.mem.map(self.ql.stack_address, self.ql.stack_size)
+            if self.ql.archtype== QL_X86:
+                self.ql.register(UC_X86_REG_ESP, self.ql.stack_address + 0x3000)
```
Comment (xwings) [inline qiling/os/uefi/uefi.py:157]: Hi,

we also updated this.

new register will be 

ql.reg.esp for read and ql.reg.esp = 0x11 for write

you can also use 

ql.reg.read() and ql.reg.write

sorry for the mass we've created.

**Comment by xwings** [comment]
you can also add in qiling/profile/uefi.ql, since UEFI is a os by itself

File: qiling/const.py [hunk 2]
```diff
@@ -35,6 +36,6 @@
 QL_DEBUGGER = [QL_IDAPRO, QL_GDB]
 QL_ARCH = [QL_ARM, QL_ARM64, QL_MIPS32, QL_X86, QL_X8664]
 QL_ENDINABLE = [QL_MIPS32, QL_ARM]
-QL_OS = [QL_LINUX, QL_FREEBSD, QL_MACOS, QL_WINDOWS]
+QL_OS = [QL_LINUX, QL_FREEBSD, QL_MACOS, QL_WINDOWS, QL_UEFI]
```
Comment (xwings) [inline qiling/const.py:39]: the new method will be QL_OS.UEFI you need a merge and fix all these

File: qiling/debugger/gdbserver/gdbserver.py [hunk 3]
```diff
@@ -266,14 +266,14 @@ def handle_p(subcmd):
...
-                            reg_value = self.ql.register(registers_x86[reg_index-1])
+                            reg_value = self.ql.register(self.ql.arch.get_reg_table()[reg_index-1])
                         else:
                             reg_value = 0
                         reg_value = self.ql.addr_to_str(reg_value)
                     
                     if self.ql.archtype== QL_X8664:
                         if reg_index <= 32:
-                            reg_value = self.ql.register(registers_x8664[reg_index-1])
+                            reg_value = self.ql.register(self.ql.arch.get_reg_table()[reg_index-1])
```
Comment (xwings) [inline qiling/debugger/gdbserver/gdbserver.py:276]: ql.reg.table should work for this.

again, ql.reg.read to replace ql.register

File: qiling/utils.py [hunk 4]
```diff
@@ -54,6 +55,7 @@ def ostype_convert(ostype):
         "macos": QL_MACOS,
         "freebsd": QL_FREEBSD,
         "windows": QL_WINDOWS,
+        "uefi": QL_UEFI,
```
Comment (xwings) [inline qiling/utils.py:58]: all these move to const.py

File: qiling/utils.py [hunk 5]
```diff
@@ -233,7 +235,11 @@ def ql_pe_check_archtype(path):
     arch = machine_map.get(pe.FILE_HEADER.Machine)
 
     if arch:
-        ostype = QL_WINDOWS
+        if pe.OPTIONAL_HEADER.Subsystem >= pefile.SUBSYSTEM_TYPE['IMAGE_SUBSYSTEM_EFI_APPLICATION'] and \
+        pe.OPTIONAL_HEADER.Subsystem <= pefile.SUBSYSTEM_TYPE['IMAGE_SUBSYSTEM_EFI_ROM'] :
+            ostype = QL_UEFI
```
Comment (xwings) [inline qiling/utils.py:240]: just a note: QL_OS.UEFI

**Comment by xwings** [comment]
and also add your name into credits.txt , thanks !

**Comment by xwings** [comment]
Long awaited feature! this is a wonderful job!

**Comment by xwings** [comment]
There is still some small bug, i will merge and fix it.

**Comment by aquynh** [comment]
@liba2k is there any TODO for this work?

**Comment by liba2k** [comment]
> @liba2k is there any TODO for this work?

Not all the hooks (sort of APIs) are implemented, but all the important ones are.

File: qiling/loader/pe_uefi.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+from .loader import QlLoader
+
+class QlLoaderPE_UEFI(QlLoader):
+    def __init__(self, ql):
+        super()
+        self.ql = ql
+    
+    def run(self):    
+        self.ql.tpl = 4 # TPL_APPLICATION
+        self.ql.hook_override = {}
```
Comment (aquynh) [inline qiling/loader/pe_uefi.py:34]: we have set_api() method in Qiling class, so that you may use it instead of this hook_override ?

**Comment by aquynh** [comment]
file gen_hooks.py is dead code?

**Pull request merged**